### PR TITLE
Review suggestions

### PR DESCRIPTION
**Pull request by thaJeztah**
Opening my branch with suggested changes as a pull-request (see https://github.com/docker/cli/pull/2708#pullrequestreview-591079988)

- remove fields from cpWithProgress that were not used
- only wrap cpWithProgress when printing progress
- rename cpWithProgress to copyProgressReader
- use aec for terminal control sequences
- print progress to STDERR instead of STDOUT
- don't use terminal detection to set the default, and instead perform
  it if no value was set. This prevents confusion when doing, e.g.:
    docker cp --help | grep ...
- return early if "quiet", to simplify multiple "if" statements

### CODE REVIEW

File: cli/command/container/cp.go [hunk 1]
```diff
@@ -98,14 +94,18 @@ func NewCopyCommand(dockerCli command.Cli) *cobra.Command {
 			}
 			opts.source = args[0]
 			opts.destination = args[1]
+			if !cmd.Flag("quiet").Changed {
+				// User did not specify "quiet" flag; suppress output if no terminal is attached
+				opts.quiet = !dockerCli.Out().IsTerminal()
+			}
```
Comment (thaJeztah) [inline cli/command/container/cp.go:100]: This is a bit "quirky" but fixes the `docker cp --help | grep quiet` case (which is confusing), but still allows automatic detection _if_ the user didn't explicitly set `--quiet` / `--quiet=true` / `--quiet=false`

File: cli/command/container/cp.go [hunk 2]
```diff
@@ -98,14 +94,18 @@ func NewCopyCommand(dockerCli command.Cli) *cobra.Command {
...
 			return runCopy(dockerCli, opts)
 		},
 	}
 
 	flags := cmd.Flags()
 	flags.BoolVarP(&opts.followLink, "follow-link", "L", false, "Always follow symbol link in SRC_PATH")
 	flags.BoolVarP(&opts.copyUIDGID, "archive", "a", false, "Archive mode (copy all uid/gid information)")
-	flags.BoolVarP(&opts.quiet, "quiet", "q", !dockerCli.Out().IsTerminal(), "Does not display copying process")
+	flags.BoolVarP(&opts.quiet, "quiet", "q", false, "Suppress progress output during copy. Progress output is automatically suppressed if no terminal is attached")
```
Comment (thaJeztah) [inline cli/command/container/cp.go:108]: Description is a bit lengthy now; I need to think for a shorter description 😅

File: cli/command/container/cp.go [hunk 3]
```diff
@@ -217,17 +224,16 @@ func copyFromContainer(ctx context.Context, dockerCli command.Cli, copyConfig cp
...
 
-	if !copyConfig.quiet {
-		fmt.Fprint(dockerCli.Out(), "Preparing to copy...\n")
-		fmt.Fprint(dockerCli.Out(), "\033[s")
+	if copyConfig.quiet {
+		return archive.CopyTo(preArchive, srcInfo, dstPath)
 	}
 
+	fmt.Fprint(dockerCli.Err(), "Preparing to copy...\n")
+	fmt.Fprint(dockerCli.Err(), "\033[s")
```
Comment (thaJeztah) [inline cli/command/container/cp.go]: This should probably be replaced with a `aec.XX`. I wasn't sure what code this was ☺️

File: cli/command/container/cp.go [hunk 4]
```diff
@@ -217,17 +224,16 @@ func copyFromContainer(ctx context.Context, dockerCli command.Cli, copyConfig cp
...
+	fmt.Fprint(dockerCli.Err(), aec.Up(1))
 	res := archive.CopyTo(preArchive, srcInfo, dstPath)
-
-	if !copyConfig.quiet {
-		fmt.Fprint(dockerCli.Out(), "\n")
-		fmt.Fprint(dockerCli.Out(), "Successfully copied ", units.HumanSize(float64(copiedSize)), " to ", dstPath, "\n")
-	}
+	fmt.Fprint(dockerCli.Err(), "\n")
```
Comment (thaJeztah) [inline cli/command/container/cp.go:235]: Changed all output to print on `STDERR` (perhaps needs to be discussed on the docker/cli PR)

**Comment by maximillianfx** [comment]
First of all, thanks for your comments and PR. I'll check all the points.

File: cli/command/container/cp.go (hunk 1 above)
Comment (maximillianfx) [reply cli/command/container/cp.go:100]: Good point, works beauty 😄

File: cli/command/container/cp.go (hunk 2 above)
Comment (maximillianfx) [reply cli/command/container/cp.go:108]: Oh, I think this ok, neither too longo nor too small.

File: cli/command/container/cp.go (hunk 4 above)
Comment (maximillianfx) [reply cli/command/container/cp.go:235]: Uhm, I'm confusing with this change. Why we need change the output to `STDERR`?

Comment (maximillianfx) [reply cli/command/container/cp.go:235]: Uhm, I'm confusing with this change. Why we need change the output to `STDERR`?

Comment (maximillianfx) [reply cli/command/container/cp.go:235]: Oh, we will use `STDERR` only when `quiet` is true?

File: cli/command/container/cp.go (hunk 3 above)
Comment (maximillianfx) [reply cli/command/container/cp.go]: This cursor command (`\033[s`) saves the cursor position, in this case, the beginning the new line.

**Review by maximillianfx — COMMENTED** [review]
Good points, I've made some observations and questions, thanks!

File: cli/command/container/cp.go (hunk 4 above)
Comment (thaJeztah) [reply cli/command/container/cp.go:235]: There's slightly more details in https://github.com/docker/cli/pull/1859/files. Thinking of that, perhaps the `Successfully copied` output should be printed on `STDOUT`.

The general idea is that "informational" output generally goes to `STDERR`; this allows users to redirect output elsewhere.

For example;

```console
$ curl https://get.docker.com 2> progress.log | grep Desktop
		echo "WSL DETECTED: We recommend using Docker Desktop for Windows."
		echo "Please get Docker Desktop from https://www.docker.com/products/docker-desktop"
					echo "Please get Docker Desktop from https://www.docker.com/products/docker-desktop"
```

In the above, informational output (progress bars) of curl are printed on `STDERR` and can be redirected (in this case I'm sending it to `progress.log`), but the output of the file that's downloaded is still on `STDERR`, and I can use `grep` to search in the content of that file.

Comment (maximillianfx) [reply cli/command/container/cp.go:235]: Uhm, ok, I understand that, nice.

File: cli/command/container/cp.go (hunk 3 above)
Comment (maximillianfx) [reply cli/command/container/cp.go]: I think that `aec.save()` could solve this.

Comment (maximillianfx) [reply cli/command/container/cp.go]: Hello @thaJeztah, how are you? Did you see this change in this line?

Comment (thaJeztah) [reply cli/command/container/cp.go]: oh! missed it (or thought you would include it in after this was merged); let me update

**Comment by thaJeztah** [comment]
@maximillianfx updated; PTAL

**Comment by maximillianfx** [comment]
LGTM, I'll made some tests here to see if breaks something.

**Pull request merged**